### DISCUSSION TITLE
Use abi param names with key remapping for functions with fully named inputs

### DISCUSSION THREAD

**Discussion by fubhy** [wevm/viem#369] [Idea / Feature Request] [answered]
I'm sure this has been debated before but is there a reason why functions with fully named inputs do not use typescript key remapping to generate an object type instead of an unnamed tuple?

I see that there are already utils for handling this case, but they are not used for AbiParameters. Hence, I'm curious... Why not sth. like this?

```typescript
type AbiParametersToPrimitiveTypes<TAbiParameters extends readonly AbiParameter[], TAbiParameterKind extends AbiParameterKind = AbiParameterKind> = _HasUnnamedAbiParameter<TAbiParameters> extends true ? readonly [
    ...{
        [K in keyof TAbiParameters]: AbiParameterToPrimitiveType<TAbiParameters[K], TAbiParameterKind>;
    }
] : {
    [K in TAbiParameters[number] as K extends {
        name: string
    } ? K['name'] : never]: AbiParameterToPrimitiveType<K, TAbiParameterKind>;
};
```

**Comment by fubhy** [comment]
The same could be done for outputs of course (if all outputs are named). I'm sure I'm missing sth. here :-)

**Comment by jxom** [comment, marked as answer]
Is this discussion along the lines of what you mean? https://github.com/pancakeswap/pancake-frontend/discussions/6558#discussioncomment-5568247

**Reply by jxom** [reply]
And this: https://github.com/wagmi-dev/viem/issues/346

**Reply by jxom** [reply]
Basically, introducing an object shape would introduce ambiguity between a “struct”ish tuple and a “positional” tuple – we feel like they should be treated separately to avoid confusion.

**Comment by fubhy** [comment]
Ok that makes sense. And dynamically named tuples would be dope indeed. Thanks for clarifying.

**Comment by fubhy** [comment]
I feel like this would be a great candidate for a FAQ :-P. Seems like that question comes up frequently.

**Reply by tmm** [reply]
Added!

https://github.com/wagmi-dev/viem/blob/main/.github/FAQ.md#why-are-contract-function-args-with-fully-named-inputs-represented-as-unnamed-tuple-types-instead-of-object-types

**Comment by emretepedev** [comment]
Is there any update for this feature? @tmm 
It would be nice to have a function like toObject() as in Ethers v6.

**Reply by tmm** [reply]
No update. You should implement something in userland if you want this feature.
